### PR TITLE
Pass metrics object for Scan, Timeseries and GroupBy queries during cursor creation

### PR DESCRIPTION
**Pull request by rohangarg**
Although we have now started emitting vectorized metric to the emitter, only topn query reports that metric. Currently scan, timeseries and groupby queries don't report that metric - this is change will support that reporting.

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/distinctcount/src/test/java/org/apache/druid/query/aggregation/distinctcount/DistinctCountTimeseriesQueryTest.java [hunk 1]
```diff
@@ -100,7 +101,7 @@ public void testTimeseriesWithDistinctCountAgg() throws Exception
                                   .build();
 
     final Iterable<Result<TimeseriesResultValue>> results =
-        engine.process(query, new IncrementalIndexStorageAdapter(index)).toList();
+        engine.process(query, new IncrementalIndexStorageAdapter(index), new DefaultTimeseriesQueryMetrics()).toList();
```
Comment (suneet-s) [inline extensions-contrib/distinctcount/src/test/java/org/apache/druid/query/aggregation/distinctcount/DistinctCountTimeseriesQueryTest.java:104]: nit: Can you verify returned by calling `engine.process(..)`

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 2]
```diff
@@ -449,7 +451,9 @@ public static Collection<Object[]> constructorFeeder()
           final String testName = StringUtils.format("config=%s, runner=%s, vectorize=%s", config, runner, vectorize);
 
           // Add vectorization tests for any indexes that support it.
-          if (!vectorize || QueryRunnerTestHelper.isTestRunnerVectorizable(runner)) {
+          if (!vectorize ||
+              (QueryRunnerTestHelper.isTestRunnerVectorizable(runner) &&
+               config.getDefaultStrategy().equals(GroupByStrategySelector.STRATEGY_V2))) {
```
Comment (suneet-s) [inline processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:456]: why was this change needed?

**Comment by suneet-s** [comment]
Screen Shot 2022-05-02 at 3 43 10 PM

I was able to test this change in a cluster and build some visualizations on top of the metrics being emitted. I think this is a step in the right direction and most of my comments are about passing the metrics object into more places so we can get even more visibility. +1 from me after CI passes

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategy.java [hunk 3]
```diff
@@ -187,7 +188,10 @@ Sequence<ResultRow> processSubtotalsSpec(
    *
    * @return result sequence for the storage adapter
    */
-  Sequence<ResultRow> process(GroupByQuery query, StorageAdapter storageAdapter);
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategy.java:190]: I didn't do it in those methods since they take the subquery/subtotal result sequence in the method. I think the inlined subquery should execute using a separate query object and hence should emit their own metrics object. For pushed down subqueries, I'm not sure about their metrics framework as of now but my guess is that it should use the same metrics object as the outer one. will check and update

File: processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV1.java [hunk 4]
```diff
@@ -233,7 +235,8 @@ public Sequence<ResultRow> apply(Interval interval)
                         outerQuery.withQuerySegmentSpec(
                             new MultipleIntervalSegmentSpec(ImmutableList.of(interval))
                         ),
-                        new IncrementalIndexStorageAdapter(innerQueryResultIndex)
+                        new IncrementalIndexStorageAdapter(innerQueryResultIndex),
+                        null
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV1.java:239]: I passed null here since passing the query metrics needed changes in the interface - and since this execution is synthetic over the inner query results and would always be non-vectorized (due to the storage adapter used) I skipped it. If you think, I can try to make that change

File: processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 2 above)
Comment (rohangarg) [reply processing/src/test/java/org/apache/druid/query/groupby/GroupByQueryRunnerTest.java:456]: this was needed because strategy-v1 doesn't implement vectorized cursors but we were adding vectorization tests in the earlier code. This check was leading to cases where we were running strategy-v1 with vectorize test parameter as true which led to mismatch in assertions.

**Review by suneet-s — COMMENTED** [review]
LGTM. It's a step in the right direction and will increase visibility 🥳 

Can you add this dimension to `docs/operations/metrics.md` so users know about this new dimension that is being emitted.

**Review by suneet-s — APPROVED** [review]
